sds_in_a_box/sds_in_a_box_stack.py

```diff
@@ -72,27 +83,122 @@ def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
             )
         )
 
-        # This is the role that the lambdas will assume
-        # We'll narrow things down later
+########### LAMBDA FUNCTIONS
+
+        # This is where we install dependencies for the lambda functions
+        # We take advantage of something called "lambda layers"
+        os.system("pip install requests -t ./external/python")
+        os.system("pip install python-jose -t ./external/python")
```

Comment:
why are these dependencies being installed here and not part of the library requirements?

Reply:
So these are the libraries that the lambda function needs in order to function.  Generally we handle dependencies in one of three ways in lambda functions:

1) We use Docker containers for our lambda functions, and have that dockerfile install all dependencies.  (https://docs.aws.amazon.com/lambda/latest/dg/images-create.html).  This is the route that Libera is taking when it comes to python dependencies.  *However*, what bothers me about that is we can't edit the code directly from the AWS console any more, making it much harder to develop and debug.  I really enjoy the flexibilty of making tweaks directly in the console.  

2) We can also package the dependencies into the lambda function itself.  One potentatial problem here is that the total amount of python code can't exceed 50MB, so it's possible that your code and all of the code in the libraries we need get so large that it can't fit in the lambda function any more.  

3) We use something called "lambda layers", which is what I do here.  ( https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html ) I'm installing things into a seperate directory locally, and CDK takes that directory, zips it up, and creates a "layer" object from it.  You can go here and check out the most recent layer this stack created: https://us-west-2.console.aws.amazon.com/lambda/home?region=us-west-2#/layers/SDSDependencieskugzmjqr394CC578/versions/1?tab=versions.  

Now I really have no idea if this is the best way to go about it.  I just googled how to handle python dependencies in the CDK, and it seems like its possible lambda might just automatically install and include things in a requirements.txt file.  I'll experiment a little more with this and see if there is a nicer looking way to go about this that doesn't install things into a local directory.

Reply:
So I was just playing around with lambda dependencies for awhile, and unfortunately it doesn't look like there is a good way to do this without things getting pretty complex.  

There is an experimental construct in the CDK which is supposed to be able to auto-install a the libraries in a requirements.txt file: https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_lambda_python_alpha/PythonFunction.html

But since it is experimental, it might not make sense to be using this yet.  Hopefully it'll get out of alpha soon.  

There is also a way to specify using a Docker image to help build your lambda function, as seen here: https://docs.aws.amazon.com/cdk/api/v1/docs/aws-lambda-readme.html#bundling-asset-code.  (That code is in javascript, but I sort of got it working for python.)  The problem is, this will spin up a docker container locally in order to help build the code that gets uploaded to the lambda function.  But since we're doing our development inside of a docker container already, building a Docker image inside of another Docker image looks like it requires a lot of weird parameters set, if it is even possible at all.  

So all that is to say, I don't necessarily see a better option at the moment.  I'm open to ideas though.  It might be something we have to revisit later on, maybe when development has stabilized a bit?

Reply:
Please don't system install, that is dangerous and if I'm running locally I absolutely don't want this happening for me without me knowing.

See here for an example of how we structured this for swxtrec_cdk. There are a few various lambda models over there where we put this `requirements.txt` file in and it works nicely.
https://github.com/SWxTREC/swxtrec-cdk/tree/main/swxtrec_cdk/msis/lambda

I think using the alpha Lambda version is totally fine. We may have to update syntax later, but that is easy enough to do.

Reply:
Alrighty, I'll try using the alpha version.

Reply:
Darn, so the problem with the alpha version and a requirements.txt file is that it still ultimately uses a Docker container to install the requirements for the lambda.  And then the problem there is that we're already running our code inside of a docker container, so we get an error about not being able to find Docker.  So the issue I mentioned above still applies.  

It may be possible by specifying custom bundling options as seen here: https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk.aws_lambda_python_alpha/README.html#custom-bundling

Though if we got that working, we'd have to complicate the CDK a bit by adding some sort of user input that specifies if you're running this CDK inside a docker container or not.